webpack/components/NewTemplateSync/components/TextButtonField.js

```diff
@@ -0,0 +1,212 @@
+    blank={blank}
+    item={item}
+    disabled={disabled}
+    validate={item.validate}
+    fieldRequired={fieldRequired}
+  />
+);
+
+const FieldType = ({ item, fieldSelector, input, disabled, blank }) => {
+  if (!fieldSelector) {
+    return <InputField input={input} disabled={disabled} />;
+  }
+
+  switch (fieldSelector(item)) {
+    case 'text':
+      return <InputField input={input} disabled={disabled} />;
+    case 'select':
+      return (
+        <SelectField
+          input={input}
+          blank={blank}
+          item={item}
+          disabled={disabled}
+        />
+      );
+    case 'checkbox':
+      return <CheckboxField input={input} item={item} disabled={disabled} />;
+    default:
+      throw new Error(`Unknown field type ${fieldSelector(item)} for ${item}`);
+  }
+};
+
+const InputField = ({ input, disabled }) => (
+  <FormControl {...input} type="text" disabled={disabled} />
+);
+
+InputField.propTypes = {
+  input: PropTypes.object.isRequired,
+  disabled: PropTypes.bool.isRequired,
+};
+
+const SelectField = ({ input, blank, item, disabled }) => (
+  <FormControl {...input} componentClass="select" disabled={disabled}>
+    <BlankOption blank={blank} />
+    {item.selection.map(opt => (
+      <option key={opt.value} value={opt.value}>
+        {opt.label}
+      </option>
+    ))}
+  </FormControl>
+);
+
+const CheckboxField = ({ input, item, disabled }) => (
+  <Checkbox {...input} disabled={disabled} />
+);
+
+const BlankOption = ({ blank }) => {
+  if (Object.keys(blank).length === 0) {
+    return null;
+  }
+  return (
+    <option key={blank.value} value={blank.value}>
+      {blank.label}
+    </option>
+  );
+};
+
+BlankOption.propTypes = {
+  blank: PropTypes.object.isRequired,
+};
+
+CheckboxField.propTypes = {
+  input: PropTypes.object.isRequired,
+  item: PropTypes.object.isRequired,
+  disabled: PropTypes.bool.isRequired,
+};
+
+SelectField.propTypes = {
+  input: PropTypes.object.isRequired,
+  blank: PropTypes.object.isRequired,
+  item: PropTypes.object.isRequired,
+  disabled: PropTypes.bool.isRequired,
+};
+
+TextButtonField.propTypes = {
+  item: PropTypes.object,
+  label: PropTypes.string.isRequired,
+  name: PropTypes.string.isRequired,
+  className: PropTypes.string,
+  inputClassName: PropTypes.string,
+  blank: PropTypes.shape({
+    label: PropTypes.string,
+    value: PropTypes.string,
+  }),
+  buttonAttrs: PropTypes.shape({
+    buttonText: PropTypes.node,
+    buttonAction: PropTypes.func,
+  }).isRequired,
+  fieldSelector: PropTypes.func,
+  validate: PropTypes.array,
+  disabled: PropTypes.bool,
+  fieldRequired: PropTypes.bool,
+  tooltipHelp: PropTypes.node,
+};
+
+TextButtonField.defaultProps = {
+  blank: { label: 'Choose one...', value: '' },
+  className: '',
+  inputClassName: 'col-md-6',
+  validate: [],
+  disabled: false,
+  fieldRequired: false,
+  tooltipHelp: null,
+  fieldSelector: () => 'text',
+  item: {},
+};
+
+export default TextButtonField;
```

Comment:
This file contains too many things.
The UI team has a golden rule of one component per file. I believe following this rule here will make it much more readable and easier to maintain.